partiql-parser/src/result.rs

```diff
@@ -2,38 +2,125 @@
 
-use crate::location::{ByteOffset, Position};
+use std::fmt::Debug;
 
-use crate::lalr::{Spanned, Token};
+use crate::lalr::Token;
+use crate::location::Located;
 use crate::LexError;
 use thiserror::Error;
 
 /// General [`Result`] type for the PartiQL parser.
-pub type ParserResult<'input, T> = Result<T, ParserError<'input>>;
+pub type ParserResult<'input, T, Loc> = Result<T, ParserError<'input, Loc>>;
 
 /// Errors from the PartiQL parser.
 #[derive(Error, Debug, Clone, PartialEq, Eq, Hash)]
-pub enum ParserError<'input> {
+pub enum ParserError<'input, Loc>
+where
+    Loc: Debug, // TODO this should be `Loc: Display`
+{
     /// Indicates that there was a problem with syntax.
-    #[error("Syntax Error: {message} ({position})")]
-    SyntaxError { message: String, position: Position },
+    #[error("Syntax Error: {} at [{:?}]", _0.inner, _0.location)]
+    SyntaxError(Located<String, Loc>),
 
-    /// There was a token that was ot expected
-    // TODO how to report location in errors
-    #[error("Unexpected token [{:?}] at [TODO]", token.1)]
-    UnexpectedToken {
-        /// The unexpected token of type `T` with a span given by the two `L` values.
-        token: Spanned<Token<'input>, ByteOffset>,
-        // TODO expected: ...,
-    },
+    /// There was a token that was not expected
+    #[error("Unexpected token [{:?}] at [{:?}]", _0.inner.token, _0.location)]
+    UnexpectedToken(UnexpectedToken<'input, Loc>),
 
-    #[error("{cause} ({position})")]
-    LexicalError { cause: LexError, position: Position },
+    /// There was an error lexing the input
+    #[error("{} at [{:?}]", _0.inner, _0.location)]
+    LexicalError(LexicalError<'input, Loc>),
 
     /// Indicates that there is an internal error that was not due to user input or API violation.
-    #[error("Illegal State: {message}")]
-    IllegalState { message: String },
+    #[error("Illegal State: {0}")]
+    IllegalState(String),
 }
 
+impl<'input, Loc: Debug> ParserError<'input, Loc>
+where
+    Loc: Debug, // TODO this should be `Loc: Display`
+{
+    /// Maps an `ParserError<Loc>` to `ParserError<Loc2>` by applying a function to each variant
+    pub fn map_loc<F, Loc2>(self, tx: F) -> ParserError<'input, Loc2>
+    where
+        Loc2: Debug, // TODO this should be `Loc2: Display`
+        F: Fn(Loc) -> Loc2,
+    {
+        match self {
+            ParserError::SyntaxError(l) => ParserError::SyntaxError(l.map_loc(tx)),
+            ParserError::UnexpectedToken(l) => ParserError::UnexpectedToken(l.map_loc(tx)),
+            ParserError::LexicalError(l) => ParserError::LexicalError(l.map_loc(tx)),
+            ParserError::IllegalState(s) => ParserError::IllegalState(s),
+        }
+    }
+}
+
+pub type LexicalError<'input, L> = Located<LexError<'input>, L>;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct UnexpectedTokenData<'input> {
+    /// The unexpected token
+    pub token: Token<'input>,
+    // TODO expected: ...,
```

Comment:
nit: could add an issue to track adding `expected`. And could reference at this TODO: https://github.com/partiql/partiql-lang-rust/blob/dcef86525302fdc990688b8c2220717d92d22376/partiql-parser/src/lalr/mod.rs#L91

Reply:
Added #89